### PR TITLE
Implementation of consistent boundary flux method for dynamic topography

### PR DESCRIPTION
**Pull request by ian-r-rose**
This is an implementation of the consistent-boundary-flux method for computing dynamic topography, which is supposed to be more accurate than the so-called "pressure smoothing" method (which is what we currently use). Benchmarking efforts are ongoing, but here is an initial convergence test:
Edit: fixed figure
convergence

### CODE REVIEW

**Comment by ian-r-rose** [comment]
cc @jaustermann

**Comment by gassmoeller** [comment]
Well this does not like an improvement to me :smile:. Do you ask for a review already?

**Comment by ian-r-rose** [comment]
Whoops, I switched the labels.

**Comment by gassmoeller** [comment]
That makes more sense then

**Comment by gassmoeller** [comment]
Let us know when you want a review.

**Comment by ian-r-rose** [comment]
Here is a convergence test for the case one benchmark of Zhong et al 93:
zhong93

CBF is significantly better. I believe it stops getting better at the end due to the fact that the analytical solution against which we are comparing is for a delta function at depth, whereas in the ASPECT grid the density perturbation has a finite depth.

**Comment by bangerth** [comment]
Nice. Could you tie the width of the zone to the mesh size (at least for the last data point/refinement)?

**Comment by ian-r-rose** [comment]
Yes, with a bit more work I can do that.

**Comment by ian-r-rose** [comment]
Updated the figure to tighten the thickness of the perturbation upon mesh refinement,

**Comment by ian-r-rose** [comment]
Okay, I am still working on some benchmarking, but I think I am ready for some code review.

**Comment by bangerth** [comment]
On 05/12/2017 01:07 PM, Ian Rose wrote:
> Yes, with a bit more work I can do that.

I'm only talking about the one data point to see whether it makes a difference.

**Comment by bangerth** [comment]
Nice --- master is of course also significantly better, but that's still a nice improvement!

**Comment by jaustermann** [comment]
This is a comparison of CBF and master for a 'realistic' Earth model (setup after Steinberger, 2016: S40RTS based buoyancy, continents with distinct buoyancy, depth dependent vs to density scaling, depth dependent viscosity, prescribed plate velocities at top, refinement of 5 in the upper 1000km and 3 below that). The CBF picks up a stronger signal where there is a lot of flow / cooling very close to the surface, e.g. at plate boundaries and ocean lithosphere cooling (the difference there is order ~10%). This seems right to me since I would assume to run into resolution issues with the 'master method' at these locations and the CBF seems to be doing better at the same resolution. The difference is zero in the continents, which is probably due to the constant buoyancy that is set within the continental lithosphere. That means there is little flow within the much thicker continental lithosphere (as it should be) and the source of the dynamic topography is deeper, which is picked up equally well by the CBF and current (master) method. 

diff-01

**Comment by ian-r-rose** [comment]
Another test, this time using the dynamic topography solution for @cedrict's annulus benchmark.
annulus_convergence
Fitting a line through the convergence finds first order for the pressure smoothing method, second order for CBF.

**Comment by bangerth** [comment]
Very nice!

File: source/postprocess/dynamic_topography.cc [hunk 1]
```diff
@@ -97,160 +257,190 @@ namespace aspect
...
+    /**
+     * Register the other postprocessor that we need: BoundaryPressures
+     */
+    template <int dim>
+    std::list<std::string>
+    DynamicTopography<dim>::required_other_postprocessors() const
+    {
+      std::list<std::string> deps;
+      deps.push_back("boundary pressures");
+      return deps;
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: you can probably simplify this as:
```
return std::list<std::string> (1,"boundary pressures");
```

File: source/postprocess/dynamic_topography.cc [hunk 2]
```diff
@@ -97,160 +257,190 @@ namespace aspect
...
-                                              Utilities::MPI::sum (integrated_lowersurface_topography,this->get_mpi_communicator())
-                                              / Utilities::MPI::sum (integrated_lowersurface_area,this->get_mpi_communicator()):
-                                              0;
+    /**
+     * Output the dynamic topography solution to
+     * a file.
+     */
+    template <int dim>
+    void
+    DynamicTopography<dim>::output_to_file(bool upper, std::vector<std::pair<Point<dim>, double> > &stored_values)
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: make the first argument a `const bool`(may have been wrong before)

Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: and break the line please

File: source/postprocess/dynamic_topography.cc [hunk 3]
```diff
@@ -97,160 +257,190 @@ namespace aspect
...
+    {
+      std::ostringstream output;
 
-      // Write the solution to an output file
-      // if (DT_mean_switch == true) subtract the average dynamic topography,
-      // otherwise leave as is
-      for (unsigned int i=0; i<stored_values_upper.size(); ++i)
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: Please keep this comment

File: source/postprocess/dynamic_topography.cc [hunk 4]
```diff
@@ -294,105 +484,23 @@ namespace aspect
...
 
+
+    /**
+     * Declare the parameters for the postprocessor.
+     */
     template <int dim>
     void
-    DynamicTopography<dim>::
-    declare_parameters (ParameterHandler &prm)
+    DynamicTopography<dim>::declare_parameters (ParameterHandler &prm)
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: please break the line

File: source/postprocess/dynamic_topography.cc [hunk 5]
```diff
@@ -294,105 +484,23 @@ namespace aspect
...
       {
         prm.enter_subsection("Dynamic Topography");
         {
-          prm.declare_entry ("Subtract mean of dynamic topography", "false",
-                             Patterns::Bool (),
-                             "Option to remove the mean dynamic topography "
-                             "in the visualization. The mean dynamic topography is calculated "
-                             "and subtracted separately for the upper and lower boundary. "
-                             "'true' subtracts the mean, 'false' leaves "
-                             "the calculated dynamic topography as is.");
```
Comment (gassmoeller) [inline source/postprocess/dynamic_topography.cc]: What happened to this option? Is that impossible for CBF, or is it automatically enabled?

File: benchmarks/annulus/annulus.cc [hunk 6]
```diff
@@ -1,8 +1,10 @@
+#include <aspect/simulator.h>
```
Comment (gassmoeller) [inline benchmarks/annulus/annulus.cc]: Is it necessary to have the changes to annulus.cc in this PR? Could you separate it into its own PR?

Comment (gassmoeller) [inline benchmarks/annulus/annulus.cc]: Ah so these are essentially modifications to check the dynamic topography for the annulus? Then lets keep them

File: benchmarks/hollow_sphere/hollow_sphere.cc [hunk 7]
```diff
@@ -650,17 +702,87 @@ namespace aspect
...
+    /**
+     * Register the other postprocessor that we need: DynamicTopography
+     */
+    template <int dim>
+    std::list<std::string>
+    HollowSpherePostprocessor<dim>::required_other_postprocessors() const
+    {
+      std::list<std::string> deps;
+      deps.push_back("dynamic topography");
+      return deps;
```
Comment (gassmoeller) [inline benchmarks/hollow_sphere/hollow_sphere.cc]: simplify this the same way is suggested for dynamic_topography.cc below

File: benchmarks/annulus/annulus.cc [hunk 8]
```diff
@@ -520,17 +548,86 @@ namespace aspect
...
+    /**
+     * Register the other postprocessor that we need: DynamicTopography
+     */
+    template <int dim>
+    std::list<std::string>
+    AnnulusPostprocessor<dim>::required_other_postprocessors() const
+    {
+      std::list<std::string> deps;
+      deps.push_back("dynamic topography");
+      return deps;
```
Comment (gassmoeller) [inline benchmarks/annulus/annulus.cc]: simplify this the same way I mentioned for dynamic_topography.cc below.

File: benchmarks/hollow_sphere/hollow_sphere.prm [hunk 9]
```diff
@@ -79,7 +79,7 @@ end
 
 subsection Mesh refinement
   set Initial adaptive refinement        = 0
-  set Initial global refinement          = 2
+  set Initial global refinement          = 4
```
Comment (gassmoeller) [inline benchmarks/hollow_sphere/hollow_sphere.prm]: why is this needed? How expensive will this be?

File: include/aspect/postprocess/dynamic_topography.h [hunk 10]
```diff
@@ -47,42 +47,79 @@ namespace aspect
         std::pair<std::string,std::string>
         execute (TableHandler &statistics);
 
+        /**
+         * Return the topography vector as calculated by CBF formulation
```
Comment (gassmoeller) [inline include/aspect/postprocess/dynamic_topography.h]: please explain what is different compared to the cellwise topography (i.e. this is the solution vector at the support points, not cellwise).

File: include/aspect/postprocess/dynamic_topography.h [hunk 11]
```diff
@@ -47,42 +47,79 @@ namespace aspect
...
+         * A vector which stores the surface stress values calculated
+         * at the midpoint of each surface cell face. This can be
+         * given to a visualization postprocessor to output dynamic topography.
          */
-        bool output_lower_surface_topography;
+        Vector<float> visualization_values;
 
         /**
-         * A parameter allows users to set the density value outside the surface
+         * A parameter allows users to set the density value
```
Comment (gassmoeller) [inline include/aspect/postprocess/dynamic_topography.h]: A parameter *that* allows ...

File: include/aspect/postprocess/dynamic_topography.h [hunk 12]
```diff
@@ -47,42 +47,79 @@ namespace aspect
...
+         * above the top surface.
          */
         double density_above;
 
         /**
-         * A parameter allows users to set the density value below the surface
+         * A parameter allows users to set the density value
```
Comment (gassmoeller) [inline include/aspect/postprocess/dynamic_topography.h]: Same here. Not your fault, but lets get it fixed, it was nearly copy-pasted into the geoid postprocessor.

**Comment by ian-r-rose** [comment]
The last benchmark I intend to do for now: dynamic topography in the `hollow sphere benchmark`. I am pretty limited in the level of refinement I can do on my laptop, but the results are still pretty good:
hollow_sphere

File: source/postprocess/dynamic_topography.cc [hunk 13]
```diff
@@ -18,57 +18,94 @@
...
+      // Storage for shape function values for the current solution.
+      // Used for constructing the known side of the CBF system.
+      std::vector<Tensor<1,dim> > phi_u (dofs_per_cell);
+      std::vector<SymmetricTensor<2,dim> > epsilon_phi_u (dofs_per_cell);
+      std::vector<double> div_phi_u (dofs_per_cell);
+      std::vector<double> div_solution(n_q_points);
+
+      // Vectors for solving CBF system.
+      Vector<double> local_vector(dofs_per_cell);
+      Vector<double> local_mass_matrix(dofs_per_cell);
```
Comment (ian-r-rose) [inline source/postprocess/dynamic_topography.cc]: The mass matrix can be stored in a vector because it is a diagonal matrix.

File: source/postprocess/dynamic_topography.cc (hunk 3 above)
Comment (ian-r-rose) [inline source/postprocess/dynamic_topography.cc]: This comment is no longer accurate.

File: source/postprocess/dynamic_topography.cc (hunk 5 above)
Comment (ian-r-rose) [inline source/postprocess/dynamic_topography.cc]: The intention is that it is no longer necessary with a more accurate dynamic topography solution. The `Subtract mean of dynamic topography` option was a bit of a workaround for poor accuracy in the pressure and stress.

File: source/postprocess/dynamic_topography.cc [hunk 14]
```diff
@@ -79,7 +116,130 @@ namespace aspect
...
+          if ( mass_matrix[global_index] > 1.e-15)
+            distributed_topo_vector[global_index] = rhs_vector[global_index]/mass_matrix[global_index];
+        }
+      distributed_topo_vector.compress(VectorOperation::insert);
+      topo_vector = distributed_topo_vector;
+
+      // Now loop over the cells again and solve for the dynamic topography.
+      // We solve for it on the support points of the system, since it can be
+      // directly put into a system vector of the right size.
+      std::vector< Point<dim-1> > face_support_points = this->get_fe().base_element( this->introspection().base_elements.temperature ).get_unit_face_support_points();
```
Comment (ian-r-rose) [inline source/postprocess/dynamic_topography.cc]: There is no problem as long as they are interpolatable onto each other. I have verified that it works fine with linear temperature elements.

File: source/postprocess/dynamic_topography.cc [hunk 15]
```diff
@@ -97,160 +257,190 @@ namespace aspect
...
 
+              fe_support_values[this->introspection().extractors.velocities].get_function_values( topo_vector, stress_support_values );
+              cell->face(face_idx)->get_dof_indices (face_dof_indices);
+              for ( unsigned int i = 0; i < face_dof_indices.size(); ++i)
+                {
+                  // Given the face dof, we get the component and overall cell dof index.
+                  const std::pair<unsigned int, unsigned int> component_index = this->get_fe().face_system_to_component_index(i);
+                  const unsigned int component = component_index.first;
+                  const unsigned int support_index = component_index.second;
+                  // Dynamic topography is stored in the temperature component.
```
Comment (ian-r-rose) [inline source/postprocess/dynamic_topography.cc]: Temperature is a convenient scalar finite element space to store the (scalar) dynamic topography.

File: benchmarks/hollow_sphere/hollow_sphere.prm (hunk 9 above)
Comment (ian-r-rose) [inline benchmarks/hollow_sphere/hollow_sphere.prm]: Accidentally changed, I will revert it.

**Comment by ian-r-rose** [comment]
Okay, review comments have been addressed.

**Comment by bangerth** [comment]
There are a few tests that fail, some because they use invalid parameters. Can you take a look?

Also, can you eventually (or now) squash the commits into a smaller number of self-contained ones? (E.g., all code changes, all tests, the changelog each into one.)

**Comment by ian-r-rose** [comment]
Yes, I broke the backwards advection, and need to fix that. Afterwards I will update the test results.

**Comment by ian-r-rose** [comment]
Okay, this should be close to ready, once I get a patch from the tester. @gassmoeller want to go another round with review?

cc @jaustermann

File: benchmarks/annulus/annulus.cc (hunk 6 above)
Comment (ian-r-rose) [inline benchmarks/annulus/annulus.cc]: Yes, they are modifications to add dynamic topography.

**Comment by gassmoeller** [comment]
Ready to go if the tester is happy.
